spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/security/oauth2/OAuth2Properties.java

```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright 2012-2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.boot.autoconfigure.security.oauth2;
+
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+import org.springframework.boot.context.properties.ConfigurationProperties;
+import org.springframework.security.oauth2.client.registration.ClientRegistrationProperties;
+
+/**
+ * Configuration properties for Spring Security OAuth 2.0 / OpenID Connect 1.0 support.
+ *
+ * @author Joe Grandja
+ * @since 2.0.0
+ */
+@ConfigurationProperties(prefix = "security.oauth2")
```

Comment:
We're moving the security configuration to `spring.security.*` in 2.0. I don't know the scope of the new OAuth2 support but if we have a client support and a resource server support it would be worth splitting things up a bit more.

This was discussed in prior reviews but `spring.security.oauth2.client.providers` sounds a better name to me. Right now we have no way to define any other client-related property.

Reply:
I'll change the prefix to `spring.security.oauth2`.

The scope for 5.0 is strictly Client support and soon after around Mar-Apr 2018 we will bring Resource Server support in 5.1. The idea was to initially house the client properties in `OAuth2Properties` and then leverage the same for Resource Server properties. Hence the placement of it in the package `org.springframework.boot.autoconfigure.security.oauth2` and the prefix of `spring.security.oauth2`.

I don't feel `spring.security.oauth2.client.providers` makes sense given that the defined [Roles](https://tools.ietf.org/html/rfc6749#section-1.1), as per spec: Resource Server, Client, Authorization Server and Resource Owner. The providers define the Resource Server and Authorization Server whereas the properties we are defining here represent the Client role and the attributes of a [Client Registration](https://tools.ietf.org/html/rfc6749#section-2). Therefore, `providers` may confuse the user since these are client-specific properties.

Reply:
`providers` is just a proposal, that's not important. What is important is that `Map` is a catch all and we may want to define _other client-specific_ properties in the near future so we can't "steal" the `client` namespace with the map. 

In other words, it should be `spring.security.oauth2.client.xyz` so that we can reuse the `spring.security.oauth2.client` namespace for other client-specific settings.

Reply:
I initially had the prefix and therefore the namespace as `spring.security.oauth2.client` but then it didn't quite read right because given the `Map<String, ClientRegistrationProperties> clients` the binding path would be `spring.security.oauth2.client.clients`. The namespace `spring.security.oauth2.client` represents a singular form whereas Spring Security allows for multiple client configurations.

As far as defining other client properties that are Boot specific, then it might make sense to have Boot's own representation of `ClientRegistrationProperties` that essentially duplicates the one in Spring Security but at least provides the flexibility to create new properties per client that are Boot specific. What do you think?

Reply:
Was just chatting with @rwinch and yes it does make sense to split up `OAuth2Properties` to `OAuth2ClientProperties` and `OAuth2ResourceServerProperties` (future). This will prevent class loading issues as both the client and resource-server modules are optional.

I'll rename `OAuth2Properties` to `OAuth2ClientProperties` for now.